### ISSUE TITLE
Online/temporal learning

### ISSUE THREAD

**Issue by tbreloff** [dmlc/MXNet.jl#29]
Is it possible to update a model fit one data point at a time?  More importantly, in what capacity is real-time time series analysis supported?  Most/all of the examples are focused on static datasets with batch fitting, but I'm not sure if that's the only way to do it with MXNet.

Some topics that I'm interested in:
- LSTM-g
- HTM (Heirarchical Temporal Memory)
- Spiking RNNs
- Reservoir Computing (Liquid State Machines)

with applications in:
- Video sequence analysis
- Auditory processing
- Semantic analysis
- Data stream sequencing, non-linear econometrics

To what extent are any of these possible through MXNet?  (Even if it's not possible today, would the framework support them naturally?)  I would consider contributing some algorithms here, or creating standalone packages that could be used by MXNet or my experimental package: [OnlineAI.jl](https://github.com/tbreloff/OnlineAI.jl).  If the MXNet framework does not support these topics naturally, I'll probably just continue to develop my OnlineAI framework independently.

**Comment by pluskid** [comment]
Short answer: Yes! The underlying symbolic API and dependency scheduler for computation is very flexible that it is not necessarily tied to a specific computation paradigm. So I think the main job will be building a high-level online learning API on top of the low level computation APIs.

Long answer: yes! And maybe my understanding of online learning is not very aligned with what you are referring to. Because what I see for the current deep learning training procedure is essentially online instead of batch. Each time a mini-batch of data is received and the parameter is tuned based on that min-batch. If you let the batch size be 1, then it is completely online. Of course, after you go over the data for a full pass (1 epoch), we repeat the whole process again for multiple epochs, making it not exactly online learning any more. But in principle, you can, for example, write a data provider that provide infinite stream of fresh data samples, so the algorithm could run forever in an online setting.

In particular, a general LSTM/RNN interface (instead of the [current example that uses explicit unrolling](http://mxnetjl.readthedocs.org/en/latest/tutorial/char-lstm.html)) is definite one thing that we are targeting.

**Comment by Andy-P** [comment]
@tbreloff I too am looking forward to a general LSTM/RNN interface. In the meantime, you might try RecurrentNN.jl (pure Julia no GPU). I use it to do online learning of LSTMs.

**Comment by tbreloff** [comment]
Thanks @Andy-P I haven't looked at this package yet.  Just curious... does this handle LSTM-g?  I'm particularly interested in building "deep LSTM networks", which I haven't actually seen much research in, and from what I have read, the LSTM-g (generalized long short term memory) framework may enable some interesting architectures.

**Comment by Andy-P** [comment]
If by "deep" you mean multi-layer, then if definitely does that. But is does not have LSTM-g as a model yet. Looking at [this paper](http://www.overcomplete.net/papers/nn2012.pdf) it seems simple enough to implement.

But it doesn't have GPU support. Which is why I want to put my effort into this project. With a proper RNN layer and custom operation layers it might be truly great!

**Comment by tbreloff** [comment]
> If by "deep" you mean multi-layer, then if definitely does that.

Cool!  Do you have any issues with backpropagating errors through the gating mechanics of the LSTM?  Do you have any examples of many-layered (ideally hierarchical) LSTM networks that were successfully trained?

> Which is why I want to put my effort into this project.

Makes sense.  I have similar goals in [OnlineAI.jl](https://github.com/tbreloff/OnlineAI.jl).  I'm currently thinking through whether I want to reorganize my code and contribute any novel pieces to other packages.  I may end up extracting the Spiking NNs into a separate package and attempting to move the more vanilla ANN stuff and general fitting strategies somewhere else.  MXNet is a possible candidate, but I'm honestly not thrilled about the language dependency... I'd prefer to have a pure-Julia implementation.

If you're interested in any collaboration on RecurrentNN.jl, please let me know.

**Comment by Evizero** [comment]
A little off-topic but to weigh in on this:

:+1: for reorganizing OnlineAI. It seems a little like a blob package atm. For example the spiking neuron stuff would make sense in it's own package (similar to NEST for example). Then again it might be more fruitful if such an reorganization is done after we got a more complete picture of LearnBase.

Concerning neural networks: MXNet seems like a great bet for that. I don't know to what extend we shut the windows community out by using binary dependencies (I am a linux user), but for research and serious application of deep learning such a compromise would be ok I think. It doesn't seem like there is currently a clean way to fully utilize the GPU for deep learning without some custom C/C++ code, but who knows what JuliaGPU comes up with.

For teaching etc it would be nicer to have a maintained pure-julia and hackable NN package with idiomatic and readable code. I am saying this because my supervisor intends to offer a class next semester on "ML with Julia", so I can definitely reserve some of my time to help in that regard. Maybe thats a role that OnlineAI could play @tbreloff? It would make sense since we both work on LearnBase

**Comment by tbreloff** [comment]
>  It seems a little like a blob package atm.

Yes certainly true.  When I started it, I didn't really intend for it to be a package others used... mostly just for my own experimentation.

> but who knows what JuliaGPU comes up with

I have high hopes that there will be something worthwhile, and that eventually a good pure-Julia implementation will be able to run (at least partially) on GPU.  Of course not all problems should run on GPU.  There is a cost and latency to GPU usage... it's only the huge networks with millions of parameters that see significant enough improvements.  I'd rather focus on making smarter networks and more efficient training algorithms, and spend time on pure speed sometime in the future.

> my supervisor intends to offer a class next semester on "ML with Julia"

Cool!  I'm really curious what this will look like.

Regarding LearnBase... I'm still trying to wrap my head around how the many types of neural nets can fit into the same framework as simpler statistics and ML (of course without too much complexity).  I think it will reveal itself with time though.

**Comment by Evizero** [comment]
> Cool! I'm really curious what this will look like

I guess the main thing to state here is that it won't be a data science class. So it's not about using Julia packages for applied ML. It's more like using the simplicity of Julia to teach ML concepts to computer scientists, since it is easy to mimic the typical ML notation in terms simple Julia code. I also intend to make the whole lecture materials open source

but this is a discussion for another time in a more appropriate place.
